src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,280 @@
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.flag.FeatureFlagSet;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.enchantment.EnchantmentEffectComponents;
+import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import net.minecraft.world.level.ItemLike;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Immutable combination of an {@link Item} and data components.
+ * Similar to an {@link ItemStack}, but immutable and without a count.
+ */
+public final class ItemResource implements IDataComponentHolderResource<Item> {
+    /**
+     * Codec for an item resource.
+     * Same format as {@link ItemStack#SINGLE_ITEM_CODEC}.
+     * Does <b>not</b> accept empty resources.
+     */
+    public static final Codec<ItemResource> CODEC = Codec.lazyInitialized(() -> ItemStack.SINGLE_ITEM_CODEC.xmap(ItemResource::of, ItemResource::toStack));
+    /**
+     * Codec for an item resource. Same format as {@link #CODEC}, and also accepts empty resources.
+     */
+    public static final Codec<ItemResource> OPTIONAL_CODEC = Codec.lazyInitialized(() -> ExtraCodecs.optionalEmptyMap(CODEC).xmap(ItemResource::fromOptional, ItemResource::asOptional));
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static ItemResource fromOptional(Optional<ItemResource> optional) {
+        return optional.orElse(ItemResource.EMPTY);
+    }
+
+    public Optional<ItemResource> asOptional() {
+        return isEmpty() ? Optional.empty() : Optional.of(this);
+    }
+
+    /**
+     * Stream codec for an item resource. Accepts empty resources.
+     */
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemResource> STREAM_CODEC = StreamCodec.composite(
+            ByteBufCodecs.holderRegistry(Registries.ITEM), ItemResource::getHolder,
+            DataComponentPatch.STREAM_CODEC, ItemResource::getComponentsPatch,
+            ItemResource::of);
+
+    public static ItemStack itemStackOf(IResourceStack<ItemResource> resourceStack) {
+        return resourceStack.resource().toStack(resourceStack.amount());
+    }
+
+    public static final ItemResource EMPTY = new ItemResource(ItemStack.EMPTY);
+    public static final ResourceStack<ItemResource> EMPTY_STACK = ResourceStack.of(EMPTY, 0);
+    public static final MutableResourceStack<ItemResource> EMPTY_MUTABLE_STACK = MutableResourceStack.of(EMPTY, 0);
+
+    /**
+     * This is used only for registry, you should not use this method!
+     */
+    @ApiStatus.Internal
+    public static ItemResource invalidateDefault(ItemLike item) {
+        if (item.asItem() == Items.AIR) return EMPTY;
+        return new ItemResource(item.asItem().getDefaultInstance().copyWithCount(1));
```

Comment:
Putting a comment here, mostly as I need to think about this more... As there are cases where this might not be what people want. While I understand the concept of vanilla's `getDefaultInstance`, I still heavily question usages of it in generic modded contexts in place of just using `new ItemStack(item)`. For example, you return this as the default resource if there is no patched in components, but for vanilla's two overrides, the stack returned by `getDefaultInstance` *does* have non default components. For potions it is water, as opposed to the default of it being empty; and for tipped arrows it is poison instead of no potion.